### ISSUE TITLE
Change HELO greeting

### ISSUE THREAD

**Issue by HeresJonny** [k9mail/k-9#4428]
Please search to check for an existing issue (including closed issues, for which the fix may not have yet been released) before opening a new issue: https://github.com/k9mail/k-9/issues?q=is%3Aissue

### Expected behavior
I would like to suggest to change the current HELO greeting from "127.0.0.1" to something else, maybe the software name/version or the name of the smartphone. Smartphone name would probably require additional permissions to read it. Another option could be to use the LAN IP while connected to a WIFI network and use public IP in other cases.

### Actual behavior
Currently, when sending email from the smartphone, K-Mail may identify as "127.0.0.1". This kind of greeting is also used by malicious software like SPAM software. Therefore this kind of HELO is currently being recommended to be blocked on connection by some mail server software like hmailserver, (cf http://www.hmailserver.com/forum/viewtopic.php?f=9&t=34659 ).

### Steps to reproduce
1. Send email from K9-Mail to your mail server software
2. Check HELO greeting in logfile
3.

### Environment
K-9 Mail version: 5.703

Android version: 9

Account type (IMAP, POP3, WebDAV/Exchange):
SMTP

Please take some time to [retrieve logs](https://github.com/k9mail/k-9/wiki/LoggingErrors) and attach them here:

**Comment by yoshimo** [comment]
You don't want your device name in there and neither your mail client. What is both accepted and generic enough to not reveal anything that can be fingerprinted?

**Comment by HeresJonny** [comment]
You could always use "localhost" instead of the IP, or something like "local.domain"... until the spammers catch on to that. You could also try to read a domain qualifier from the network. When used in LAN, the router will often add a domain qualifier like ".local" or ".fritz.box".

**Comment by SorenRR** [comment]
RFC's are made for a reason. Rightly ANY HELO/EHLO greeting not being a FQDN or an address-literal should be rejected by the server.

"localhost" or "127.0.0.1" is a clear winner in almost all SPAM/BOT databases, so have fun.

https://tools.ietf.org/html/rfc5321#section-4.1.1.1

**Comment by yoshimo** [comment]
I am not familiar enough with ALL rfcs so i do not know how to satisfy a both requirements rfc&privacy at the same time. Maybe there is something better that doesn't leak my network internals and device info, therefore i asked.

**Comment by SorenRR** [comment]
Privacy? You should greet with your FQDN or your address-literal, none of the two contain information that cannot be obtained elsewhere.

If you want privacy, stay off the 'net !

I joined "The Internet" before it was called "The Internet" and I'm still here - and nobody still don't know who I am besides the people I trust ;-)

**Comment by HeresJonny** [comment]
@SorenRR Are you the same Søren from the hmailserver forum who provides those great scripts? I'm sure you'll agree that on Github we should all work together towards solutions. While your opinion is certainly valid it would be even better if you could help us with suggestions. Thanks :-)

**Comment by SorenRR** [comment]
One and the same. 

Well, my best suggestion is to follow the RFC's to the letter. Someone a lot smarter than any of us wrote them for a reason. ;-)
Also when people question compatibility, you can shut them up by stating RFC compliance.

**Comment by cketti** [comment]
SMTP (the protocol) is in the unfortunate situation that it is used for two very different use cases. Often the services are referred to as SMTP ([MTA](https://en.wikipedia.org/wiki/Message_transfer_agent) to MTA communication) and SUBMISSION ([MUA](https://en.wikipedia.org/wiki/Email_client) to [MSA](https://en.wikipedia.org/wiki/Message_submission_agent) communication). 
K-9 Mail is an email client (MUA) and its communication therefore falls into the second category. In my opinion a well-configured submission server should…
- require transport encryption to be used (TLS)
- require authentication
- not use the EHLO argument for spam checking purposes.

This is also how many email providers work.

To make this easy to implement SMTP and SUBMISSION operate on different ports. SMTP uses port 25, SUBMISSION uses port 587 or 465 (see [RFC 8314](https://tools.ietf.org/html/rfc8314)).

RFC 5321 states this about the EHLO parameter:
>The SMTP client MUST, if possible, ensure that the domain parameter to the EHLO command is a primary host name as specified for this command in Section 2.3.5.  If this is not possible (e.g., when the client's address is dynamically assigned and the client does not have an obvious name), an address literal SHOULD be substituted for the domain name.
>
>   An SMTP server MAY verify that the domain name argument in the EHLO command actually corresponds to the IP address of the client. However, if the verification fails, the server MUST NOT refuse to accept a message on that basis.  Information captured in the verification attempt is for logging and tracing purposes.  Note that this prohibition applies to the matching of the parameter to its IP address only; see Section 7.9 for a more extensive discussion of rejecting incoming connections or mail messages.

Most mobile devices don't have a domain name that resolve to their IP address. Since NAT gateways are still very common it's also not easy to find out a device's public IP address without using a third-party service. Getting the IP address of the network interface used to connect to the submission server will most likely reveal an internal IP address. We therefore opted to always send `EHLO [127.0.0.1]` or `EHLO [IPv6:::1]`. These are syntactically valid address literals.

To summarize: do whatever voodoo you want to detect spam on SMTP. But SUBMISSION is there to serve authenticated users of you service. Requiring anything from EHLO arguments besides using the correct syntax is unreasonable because of the environment (mobile) clients work in.

**Comment by SorenRR** [comment]
Perhaps someone need to re-read the relevant RFC's. And... People really need to look into their network settings as I believe both the unix and windows world have the ability to build a FQDN from local hostname and domain.

I have a mixture of Linux, Windows XP, Windows 10, Windows Server 2003R2 and Windows Server 2012R2 machines in my LAN that all have FQDN's built from the domain name assigned via DHCP and the local hostname.

But again... The RFC's are made by people a lot smarter than all of us and so far I have not seen anything to contradict that.

**Comment by HeresJonny** [comment]
So this has been fixed?

**Comment by cketti** [comment]
> So this has been fixed?

No. In the past K-9 Mail has done what you have suggested in the original post. But this was causing issues, too.
I maintain that every SUBMISSION server that denies clients because of the HELO/EHLO argument is misconfigured.

**Comment by SorenRR** [comment]
@cketti, You do know that the ESMTP "submission protocol" used by MSA's are based on SMTP and that the SMTP protocol dictate that the HELO/EHLO greeting must include the FQDN or address literal of the MSA/MUA/MTA?

It is not your job to dictate how a mail administrator configure his server as long as he follow the RFC's. If the administrator decide to reject non-FQDN EHLO greetings it is his right. And your loss!

Just my 5c...

**Comment by cketti** [comment]
> @cketti, You do know that the ESMTP "submission protocol" used by MSA's are based on SMTP and that the SMTP protocol dictate that the HELO/EHLO greeting must include the FQDN or address literal of the MSA/MUA/MTA?

I quoted the RFC above. And if you have a closer look, you'll see that K-9 Mail is using a valid address literal.

> It is not your job to dictate how a mail administrator configure his server as long as he follow the RFC's. If the administrator decide to reject non-FQDN EHLO greetings it is his right. And your loss!

Your "spam fighting measure" (which arguably is a violation of the RFC; see recent discussions on the ietf-smtp mailing list) is preventing your own (authenticated) users from using the service for no good reason. 
It can't be that hard to limit the EHLO check to port 25 and not apply it to port 587. Everybody else seems to do it that way.

**Comment by SorenRR** [comment]
Ah... You do remember that the response to EHLO is the list of options offered like STARTTLS and AUTH? Users are not authenticated at the stage of rejection.

FYI: The postfix equivalent to what I do is: reject_non_fqdn_helo_hostname = yes

Can you please show exactly which RFC I violate by rejecting a non-compliant EHLO. 
I know I *should* send back 550, I don't... I ban the sucker for a week.

Oh and I also host a webmail solution, just in case ... ;-)

**Comment by justinacolmena** [comment]
On January 25, 2020 12:10:19 PM AKST, "Søren Raagaard Rathje" <notifications@github.com> wrote:
>Ah... You do remember that the response to EHLO is the list of options
>offered like STARTTLS and AUTH? Users are not authenticated at the
>stage of rejection.
I should hope not.
>
>FYI: The postfix equivalent to what I do is:
>reject_non_fqdn_helo_hostname = yes
>
>Can you please show exactly which RFC I violate by rejecting a
>non-compliant EHLO. 
>I know I *should* send back 550, I don't... I ban the sucker for a
>week.

https://www.linode.com/community/questions/8831/postfix-reject_non_fqdn_helo_hostname

If the mail is on port 25, as it is supposed to be from another server, then that is probably a good idea.

If on the other hand there is a client trying to log in on port 587 and use STARTTLS and AUTH for relay privileges, then the reject_non_fqdn_helo_hostname option is probably unnecessary, and will almost surely prevent authenticated clients from sending mail.

It is a little confusing to me, because the postfix configuration is somewhat "mixed" between "main.cf" and the inetd-like "master.cf" as to which options apply on which open ports.

>
>Oh and I also host a webmail solution, just in case ... ;-)

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by cketti** [comment]
In general it's quite simple.

Port 25 (SMTP): Do whatever you feel is necessary.

Port 587 (SUBMISSION): Only check HELO/EHLO syntax, not content. Require authentication. Optional (but advisable): only allow sender addresses associated with the authentication credentials used.

**Comment by SorenRR** [comment]
What is the problem with having k9mail use the FQDN?

Privacy is not a valid option! If your ISP know your IP Address then NSA, GRU, FSB, CIA, MSS, KGB and Mossad know it too.

**Comment by spaetz** [comment]
I know this issue is closed, but my Android is always connecting via VPN when on mbile, so all kmail9 will ever see is a 10.0.0.1 IP address. Refusing to take mail from it would not be smart thing though. So just for the records: k9mail cannot always see a valid FQDN.

**Comment by dejab** [comment]
I love spec design.
Not the word legitimately in RFC 5735 and MUST NOT in RFC 1122

RFC 5735               Special Use IPv4 Addresses           January 2010

   127.0.0.0/8 - This block is assigned for use as the Internet host
   loopback address.  A datagram sent by a higher-level protocol to an
   address anywhere within this block loops back inside the host.  This
   is ordinarily implemented using only 127.0.0.1/32 for loopback.  As
   described in [RFC1122], Section 3.2.1.3, addresses within the entire
   127.0.0.0/8 block do not legitimately appear on any network anywhere.

RFC 1122                    INTERNET LAYER                 October 1989

3.2.1.3
            (g)  { 127, <any> }

                 Internal host loopback address.  Addresses of this form
                 MUST NOT appear outside a host.

**Comment by cketti** [comment]
None of these RFCs are relevant for this issue. K-9 Mail is not sending IP packets using 127.0.0.1 as a sender or recipient IP address.

**Comment by dejab** [comment]
Sure it does. K9 client is trying to legitimize an IP Address that is:

Not suppose to be routable.
Does not legitimately appear on any network anywhere.
And, MUST NOT appear outside a host.

**Comment by gpz1100** [comment]
The problem is there's 2 school's of thought on this.

While I agree, port 25 should have a valid and proper FQDN.  It is unreasonable to expect this of SUBMISSION type transactions. These days, nearly all webmail services (gmail, hotmail, yahoo, etc), remove or obfuscate submission client hostname or ip for privacy reasons.

It's not unreasonable to expect this in an email client either.

Perhaps its time RFC is updated to reflect the changing needs and requirements of privacy.

**Comment by SorenRR** [comment]
@gpz1100
Not sure if you run your own server, I guess not. When have you last looked at the raw mail you sent from GMail? 95% of all Hotmail I receive is SPAM - Hot Lonely/MILF/Ukrainian women (with explicit nude pictures) looking for a partner, Not seen a Yahoo mail in ages - They still exist?

I do like Google on my server, I delete the IP address of the email sender/client in the "Received:" header before it leaves the server, but that is all. Your email address is already all over the Internet. Probably 3 million people or bot's already know your email address because YOU spread it all over when signing up for crap, concerts or shop offers!

You should read this: https://spamauditor.org/best-practices/resolve-helo-identifier/
They claim this is bad "HELO [192.168.1.1]" but it is valid per RFC!

**Comment by gpz1100** [comment]
@SorenRR 

An authenticated user sending through smtp should not need (or send) HELO on submission. That mta will have it's own HELO string when connecting to the target (receiving) MTA.

The concern is not about the email address, but the HELO content and/or local fqdn/private ip being sent.

And yes, I do self host my domain's email.  All outbound content is fully sanitized.  My server however is only open to local clients or those connected over vpn. Occasionally I do send an email using the phone or tablet.  It's for that use case where this concern becomes an issue.

**Comment by gpz1100** [comment]
I see your point. In that case, HELO should be configurable or as I believe k9 does now, sets it to localhost.
